docs/dev/stream/state/state.md

```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
+
+The collection types of state support TTL on entry level: 
+separate list elements and map entries expire independently. 
+
+The behaviour of state with TTL firstly should be configured by building `StateTtlConfiguration`:
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+StateTtlConfiguration ttlConfig = StateTtlConfiguration
+    .newBuilder(Time.seconds(1))
+    .setTtlUpdateType(StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite)
+    .setStateVisibility(StateTtlConfiguration.TtlStateVisibility.NeverReturnExpired)
+    .build();
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+val ttlConfig = StateTtlConfiguration
+    .newBuilder(Time.seconds(1))
+    .setTtlUpdateType(StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite)
+    .setStateVisibility(StateTtlConfiguration.TtlStateVisibility.NeverReturnExpired)
+    .build()
+{% endhighlight %}
+</div>
+</div>
+
+It has several options to consider. 
+The first parameter of `newBuilder` method is mandatory, it is a value of time-to-live itself.
+
+The update type configures when the time-to-live of state value is prolonged (default `OnCreateAndWrite`):
+
+ - `StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite` - only on creation and write access,
+ - `StateTtlConfiguration.TtlUpdateType.OnReadAndWrite` - also on read access.
+ 
+The state visibility configures whether the expired value is returned on read access 
+if it is not cleaned up yet (default `NeverReturnExpired`):
+
+ - `StateTtlConfiguration.TtlStateVisibility.NeverReturnExpired` - expired value is never returned,
+ - `StateTtlConfiguration.TtlStateVisibility.ReturnExpiredIfNotCleanedUp` - returned if still available.
+
+The TTL can be enabled in descriptor for any type of state:
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+StateTtlConfiguration ttlConfig = StateTtlConfiguration.newBuilder(Time.seconds(1)).build();
+ValueStateDescriptor<String> stateDescriptor = new ValueStateDescriptor<>("text state", String.class);
+stateDescriptor.enableTimeToLive(ttlConfig);
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+val ttlConfig = StateTtlConfiguration.newBuilder(Time.seconds(1)).build()
+val stateDescriptor = new ValueStateDescriptor[String]("text state", classOf[String])
+stateDescriptor.enableTimeToLive(ttlConfig)
+{% endhighlight %}
+</div>
+</div>
+
+**Notes:** 
+
+- The state backends store the timestamp of last modification along with the user value, 
+which means that enabling this feature increases consumption of state storage.
+
+- As of current implementation the state storage is cleaned up of expired value 
+only on its explicit read access per key, e.g. calling `ValueState.value()`. 
+This might change in future releases, e.g. additional strategies might be added in background to speed up cleanup.
```

Comment:
In the current implementation, expired values are only removed when they are read explicitly, e.g. by calling `ValueState.value()`. Note that this means that under the current implementation if expired state is not read, it won't be removed, possibly leading to ever growing state. This might change in future releases. Additional strategies might be added that clean up expired state in the background. 

BTW, is this true given FLINK-9938?

Reply:
I think if we merge FLINK-9938 before release, the description should be updated.

Reply:
I think we still want to improve this, because FLINK-9938 is limited in a way that it cleans up only the snapshot data but not the local one used by running pipeline atm.